src/ir_passes/recomb.ml

```diff
@@ -0,0 +1,139 @@
+(* Turns rebuilt variants into casts of the same thing.
+   Also turns catchall cases into wildcard ones when the scrutinee is already a variable. *)
+
+open Ir_def
+open Ir
+open Source
+
+(* The transformation environment is trivial for now *)
+let empty_env () = ()
+
+(* The AST traversal *)
+
+(* Does a very simple thing (for now):
+ - spot `switch var {`
+ - where a variant match happens, that immediately recombines:
+    `case (#err err) { #err err }`
+ - replace with `case (#err _) { (prim cast) var }`
+ *)
+
+let rec t_exps env = List.map (t_exp env)
+
+and case env scrutinee case = match scrutinee, case with
+  (* switch v { ...; case (#err e) { #err e } } --> switch v { ...; case (#err _) { (prim cast) v } } *)
```

Comment:
```suggestion
  (* switch x { ...; case (#err y) { #err y } } --> switch x { ...; case (#err _) { (prim cast) x } } *)
```

Reply:
I'm not super thrilled about inserting unsafe casts for optimization, but maybe we can this optimization last in the pipeline.